### PR TITLE
perf(schema): avoid duplicate `get` operations

### PR DESCRIPTION
**Pull request by GalacticHypernova**
### ❓ Type of change

- [ ] 📖 Documentation (updates to the documentation, readme or JSdoc annotations)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [x] 👌 Enhancement (improving an existing functionality like performance)
- [ ] ✨ New feature (a non-breaking change that adds functionality)
- [ ] 🧹 Chore (updates to the build process or auxiliary tools and libraries)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

This PR saves the app that is gotten through the runtimeConfig $resolve async function and is used in place of the current 3 calls, resulting in a **~99.1%** increase in performance.
Benchmark results (tested with the smallest possible timeout of 0 ms which would, in real time, likely take longer than that):
```ts
async function get(name){
        await new Promise((resolve)=>{
            setTimeout(()=>{},0)
            resolve(void 0)
        })
        return {baseURL:"",buildAssetsDir:"",cdnURL:""}
    }
    const val={}
    console.time("current")
    for (let index = 0; index < 100000; index++) {
       val.baseURL=(await get('app')).baseURL
       val.buildAssetsDir=(await get('app')).buildAssetsDir
       val.cdnURL=(await get('app')).cdnURL
    }
    console.timeEnd("current")
    const app = await get('app')
    console.time("PR")
    for (let index = 0; index < 100000; index++) {
        val.baseURL=app.baseURL
        val.buildAssetsDir=app.buildAssetsDir
        val.cdnURL=app.cdnURL
     }
     console.timeEnd("PR")
```     
image

### 📝 Checklist

- [ ] I have linked an issue or discussion.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by GalacticHypernova** [comment]
I will try to get the rest handled as well, but this was the thing that caught my eye the most.

File: packages/schema/src/config/common.ts [hunk 1]
```diff
@@ -50,7 +50,10 @@ export default defineUntypedSchema({
    * It is normally not needed to configure this option.
    */
   workspaceDir: {
-    $resolve: async (val, get) => val ? resolve(await get('rootDir'), val) : await findWorkspaceDir(await get('rootDir')).catch(() => get('rootDir'))
+    $resolve: async (val, get) => {
+      const rootDir = get('rootDir')
+      return val ? resolve(await rootDir, val) : await findWorkspaceDir(await rootDir).catch(() => rootDir)
```
Comment (GalacticHypernova) [inline packages/schema/src/config/common.ts:55]: @danielroe is the catch supposed to return the promise? Is it not meant to be awaited?

Comment (danielroe) [inline packages/schema/src/config/common.ts:55]:
```suggestion
      const rootDir = await get('rootDir')
      return val ? resolve(rootDir, val) : await findWorkspaceDir(rootDir).catch(() => rootDir)
```

**Review by danielroe — APPROVED** [review]
Nice 👌

**Comment by GalacticHypernova** [comment]
Could I request this PR to not be merged until I'm done fixing all of schema? Or is it fine if it gets merged in chunks?

**Comment by danielroe** [comment]
Oh, of course.

**Comment by GalacticHypernova** [comment]
Thank you!
I'll try to get this update ready by the end of the day.

**Comment by GalacticHypernova** [comment]
Looking at the webpack schema, I don't think making it as such is necessarily the most performant:
image

Can this be changed to something like this?
```ts
vue:  {
    $resolve: async (val, get) => {
        const vue = await get('vue')
        return {
            transformAssetUrls: {
              video: 'src',
              source: 'src',
              object: 'src',
              embed: 'src'
            },
            compilerOptions: vue.compilerOptions,
            propsDestructure: Boolean(vue.propsDestructure),
            defineModel: Boolean(vue.defineModel)
        }
    }
}
```
Or would it have some kind of side effects? Because the same thing is done with the runtimeConfig:
image
It provides it from the app already so it doesn't have to get everything separately.

File: packages/schema/src/config/webpack.ts [hunk 2]
```diff
@@ -197,15 +197,20 @@ export default defineUntypedSchema({
...
+              video: 'src',
+              source: 'src',
+              object: 'src',
+              embed: 'src'
+            },
+            compilerOptions: vue.compilerOptions,
+            propsDestructure: Boolean(vue.propsDestructure),
+            defineModel: Boolean(vue.defineModel)
+          })
+        }
```
Comment (GalacticHypernova) [inline packages/schema/src/config/webpack.ts:213]: @danielroe would a change like this make sense?

Comment (GalacticHypernova) [reply packages/schema/src/config/webpack.ts:213]: I can also just wrap everything in a big defu object, like so:
```ts
export default defineUntypedSchema({
    webpack: {
        $resolve: async (val, get) => {
            const webpack = await get('webpack')
            return defu(val, { /* all the properties in here */ })
        }
    }
})
```
Do you think that would be overkill?

Comment (GalacticHypernova) [reply packages/schema/src/config/webpack.ts:213]: Alright, I'll revert it then. Do you have any guidelines for when it's better to make it like this? (For example, the app section in common.ts is resolved the same way as the webpack change I suggested)

Comment (GalacticHypernova) [reply packages/schema/src/config/webpack.ts]: I have reverted the update, which means I *think* it is ready for review.

**Review by danielroe — APPROVED** [review]

File: packages/schema/src/config/common.ts [hunk 3]
```diff
@@ -342,15 +351,19 @@ export default defineUntypedSchema({
...
-      '**/*.{spec,test}.{js,cts,mts,ts,jsx,tsx}', // ignore tests
-      '**/*.d.{cts,mts,ts}', // ignore type declarations
-      '**/.{pnpm-store,vercel,netlify,output,git,cache,data}',
-      relative(await get('rootDir'), await get('analyzeDir')),
-      relative(await get('rootDir'), await get('buildDir')),
-      await get('ignorePrefix') && `**/${await get('ignorePrefix')}*.*`
-    ].concat(val).filter(Boolean)
+    $resolve: async (val, get) => {
+      const rootDir = await get('rootDir')
+      const ignorePrefix = await get('ignorePrefix')
```
Comment (pi0) [inline packages/schema/src/config/common.ts:356]: Small perf we could use parallel promises: `const[rootDir, ignorePrefix] = await Promise.all(get('rootDir'), get('ignorePrefix'))`

Comment (GalacticHypernova) [reply packages/schema/src/config/common.ts:356]: Oh you're right! Thanks, I missed that. I know another place this can be added in.

Comment (GalacticHypernova) [reply packages/schema/src/config/common.ts:356]: Is there a way to "append" commits to a merged PR or should I open a new one?

Comment (pi0) [reply packages/schema/src/config/common.ts:356]: You probably need a new one..

**Pull request merged**